ephyviewer/datasource/epochs.py

```diff
@@ -54,152 +55,219 @@ class WritableEpochSource(InMemoryEpochSource):
-        self.possible_labels = possible_labels
+        # put the epochs into a canonical order after loading
+        self._clean_and_set(self.all[0]['time'], self.all[0]['duration'], self.all[0]['label'], self.all[0]['id'])
         
         if color_labels is None:
-            n = len(possible_labels)
+            n = len(self.possible_labels)
             cmap = matplotlib.cm.get_cmap('Dark2' , n)
             color_labels = [ matplotlib.colors.ColorConverter().to_rgb(cmap(i)) for i in  range(n)]
             color_labels = (np.array(color_labels)*255).astype(int)
             color_labels = color_labels.tolist()
         self.color_labels = color_labels
         self.label_to_color = dict(zip(self.possible_labels, self.color_labels))
         
+    @property
+    def ep_times(self):
+        return self.all[0]['time']
+    
+    @ep_times.setter
+    def ep_times(self, arr):
+        self.all[0]['time'] = arr
+
+    @property
+    def ep_durations(self):
+        return self.all[0]['duration']
+    
+    @ep_durations.setter
+    def ep_durations(self, arr):
+        self.all[0]['duration'] = arr
+
+    @property
+    def ep_labels(self):
+        return self.all[0]['label']
+    
+    @ep_labels.setter
+    def ep_labels(self, arr):
+        self.all[0]['label'] = arr
+
+    @property
+    def ep_ids(self):
+        return self.all[0]['id']
+    
+    @ep_ids.setter
+    def ep_ids(self, arr):
+        self.all[0]['id'] = arr
+    
+    @property
+    def ep_stops(self):
+        return self.ep_times + self.ep_durations
+
+    @property
+    def id_to_ind(self):
+        return dict((id,ind) for ind,id in enumerate(self.ep_ids))
+    
     def get_chunk(self, chan=0,  i_start=None, i_stop=None):
         assert chan==0
-        return InMemoryEpochSource. get_chunk(self, chan=chan,  i_start=i_start, i_stop=i_stop)
+        ep_times = self.all[chan]['time'][i_start:i_stop]
+        ep_durations = self.all[chan]['duration'][i_start:i_stop]
+        ep_labels = self.all[chan]['label'][i_start:i_stop]
+        ep_ids = self.all[chan]['id'][i_start:i_stop]
+        return ep_times, ep_durations, ep_labels, ep_ids
+
     
     def get_chunk_by_time(self, chan=0,  t_start=None, t_stop=None):
-        #~ print(chan)
         assert chan==0
-        return InMemoryEpochSource.get_chunk_by_time(self, chan=chan,  t_start=t_start, t_stop=t_stop)
+        ep_times = self.all[chan]['time']
+        ep_durations = self.all[chan]['duration']
+        ep_labels = self.all[chan]['label']
+        ep_ids = self.all[chan]['id']
+
+        keep1 = (ep_times>=t_start) & (ep_times<t_stop) # epochs that start inside range
+        keep2 = (ep_times+ep_durations>=t_start) & (ep_times+ep_durations<t_stop) # epochs that end inside range
+        keep3 = (ep_times<=t_start) & (ep_times+ep_durations>t_stop) # epochs that span the range
+        keep = keep1 | keep2 | keep3
+
+        return ep_times[keep], ep_durations[keep], ep_labels[keep], ep_ids[keep]
```

Comment:
Why not use the parent function ?

Reply:
Because `get_chunk*` in `WritableEpochSource` needed ids, but ids were not needed in `InMemoryEpochSource`. In fact, I originally put ids in `InMemoryEpochSource` and just called the parent function in `WritableEpochSource` as you suggest, but this broke some things, as I described in 6d5987d's commit message